New Session

liwde opened the pull request:
see issue: #78

Blogpost on the Software Design: https://blog.liwde.de/posts/2018/der-ewige-kreis/ (in German)

### Architectural/Backend Changes
- [x] Completely rewrite the session service
- [x] Remove DatabaseManagerService and merge it's tasks into the session
- [x] Three parts for the new session: `local-session.ts` for local db and auth, `remote-session.ts` for remote db-connection and `session.service.ts` putting both together as Angular Service
- [x] Reworked login flow: See https://blog.liwde.de/posts/2018/der-ewige-kreis/#login-ablauf or the respective comments in the code
- [x] New Mock-Session to handle non-synced mock-db during development
- [x] Retry to connect when user was offline and goes online
- [x] Live-Sync with offline/online state change recognition
- [x] Unit Tests (some are already there, but we need more)
- [x] Documentation

liwde commented:
@sleidig @TheSlimvReal,

it would be great, if you could start looking at this. Feel free to ask, if you've got any questions!
An overview of the architecture is laid out here: https://blog.liwde.de/posts/2018/der-ewige-kreis/#redesign, with the final setup here: https://blog.liwde.de/posts/2018/der-ewige-kreis/#refactoring-und-fazit

I'm currently working on offline/online detection.

Some help regarding the tests would be much appreciated. To test the whole thing, integration tests would be nice, but they keep failing on premise (and sometimes toggle locally). If you've got ideas how to improve this, I'd be really happy to hear them :)

liwde commented:
Online/Offline-Checking now works out of the box, the UI can just grab the session and use `sessionService.getConnectionState().getStateChangeStream().subscribe()` or ``sessionService.getConnectionState().getState() === ConnectionState.connected` to get the values

sleidig commented on src/app/session/local-session.ts:
```diff
@@ -0,0 +1,130 @@
+   * Wait for the first sync of the database, returns a Promise.
+   * Resolves directly, if the database is not initial, otherwise waits for the first change of the SyncState to completed (or failed)
+   */
+  public async waitForFirstSync() {
```
Can we sketch an activity diagram to make the states entirely clear? Even a snap of a drawing would help a lot as a starting point.

liwde replied:
I added sequence diagrams and hope they help

sleidig commented on src/app/session/login-state.enum.ts:
```diff
@@ -1,31 +1,23 @@
+
+// Login at the local database
+export enum LoginState {
+  loginFailed, // wrong credentials
```
JsDoc?

sleidig commented on src/app/session/remote-session.spec.ts:
```diff
@@ -0,0 +1,16 @@
+ *
+ *     You should have received a copy of the GNU General Public License
+ *     along with ndb-core.  If not, see <http://www.gnu.org/licenses/>.
+ */
```
empty file?

sleidig commented on src/app/session/remote-session.ts:
```diff
@@ -0,0 +1,116 @@
+ */
+
+
+/**
```
move docs directly above class

liwde commented:
@sleidig Thank you for the review! I'll have a look at the code next Thursday, but I've already created some sequence diagrams detailing out what happens in which scenarios and how the three different states changed. I checked in the documentation and will add the generated SVGs as well next week -- for now there is a preview in my blog post's appendix.

liwde commented on src/app/children/child-details/child-details.component.spec.ts:
```diff
@@ -93,15 +91,14 @@ describe('ChildDetailsComponent', () => {
         PercentPipe,
-        { provide: SessionService, useValue: mockedSession},
-        DatabaseManagerService,
+        databaseServiceProvider,
```
@TheSlimvReal mentioned the same thing. But wouldn't that leave us with a cyclic dependeny, where SessionModule imports DatabaseModule for the Database, but DatabaseModule imports SessionModule for the Session to be used in the DatabaseServiceProvider?

liwde commented on src/app/session/local-session.spec.ts:
```diff
@@ -1,22 +1,16 @@
-  loggedOut,
-  loginFailed
-}
+/*
```
Yes, there are currently no tests for the local and the remote session. I'm not sure what to test here, but I'll think about it

TheSlimvReal commented on src/app/session/synced-session.service.spec.ts:
```diff
@@ -0,0 +1,278 @@
+        });
+
+        it('has the correct state after Logout', () => {
+            sessionService.logout();
```
I would sugest to log the user in before calling logout. This way the test would also pass without even calling logout thus it does not properly test the function.

TheSlimvReal commented on src/app/session/synced-session.service.spec.ts:
```diff
@@ -0,0 +1,278 @@
+    });
+
+    // These tests mock the login-methods of local and remote session.
+    // We cannot test wether the StateHandlers are in correct state, as these are set in the sub-classes themselves.
```
*whether

liwde commented:
@sleidig Any more input from your side?

TheSlimvReal commented:
I collected a list of ToDo's from the comments of @sleidig:

- [ ] Check if the DatabaseModule can be removed or refactored it with the session
- [ ] Change enum-names to be upper case
- [ ] Explain why the `pulledChanges` in the `syncStatus` is commented-out
- [ ] Reset the config.json to use the local DB on default
- [ ] Logout when being offline should be working
- [ ] Login in for the initial sync when offline should create an konstistent state

liwde commented:
> Explain why the pulledChanges in the syncStatus is commented-out

In the initial design I didn't include that sync status. Live Sync is working (tested across different browsers) and a sync completion message is shown. Is that enough or do we need the additional message? (We could add it under the `case SyncState.STARTED`.)

liwde commented:
Hi everyone,
I'm back and addressed the open issues. Please have a look again, but the issues from the last reviews were resolved :)

TheSlimvReal commented on src/app/session/mock-session.service.ts:
```diff
@@ -0,0 +1,73 @@
+        return this.database;
+    }
+    public login(username, password): Promise<LoginState> {
+        if (username === 'demo' && password === 'pass') {
```
It would be good if the login method uses the users of the mock BD. When working with the admin panel or creating new users it would be helpful if these things can also be tested in the temporary database.

liwde replied:
I'll have a look at this. As far as I remember, this piece was adapted from the old session/Database-Manager, but what you say makes sense.

sleidig replied:
you could even consider some inheritance from LocalSession (`class MockSession extends LocalSession`) and override the diverting parts

liwde replied:
The `MockSession` must (at least indirectly) extend the abstract `Session` for Dependency Injection. I'd rather not make `LocalSession` extend `Session`, as we'd need to implement the additional APIs then that don't make sense at the `LocalSession`...

liwde replied:
@TheSlimvReal The demo user is now pulled from the database instead of being hardcoded. However, it is still defined in the MockSession – should we move it to the Demo Data generation?

liwde replied:
@sleidig This is my first time working with demo data generation, and I really like it! It feels much better to have the demo user in there (see https://github.com/Aam-Digital/ndb-core/pull/246/commits/9ede8b53a42de38a790641a9a55fd6ff2524454f)

sleidig replied:
:-)

liwde commented:
At least in Firefox, indexeddb (which pouchdb heavily uses) is disabled – when I tried it, that message was actually shown at login. I didn't try Chrome, though.
You can reset your data in the devtools, it's mostly just clearing the indexeddb. Chrome even has a simple option in the devtools to delete everything of a webapp at once.

TheSlimvReal commented:
Okay when I start the app in chrome and disable the internet before login, the error message is correct. However, after turning on the internet again, the app starts trying to load the db immediately (without pressing login button) and the popup stays for ever.

liwde commented:
> However, after turning on the internet again, the app starts trying to load the db immediately (without pressing login button) and the popup stays for ever.

This is (at least in part) as designed: When you are offline, the login will automatically retry in the background. It does not make a difference, whether you are online or not. However, the popup should not stay forever. I'll have a look.

sleidig commented:
from a UX perspective I think it would be best to start the retry in background only after a successful (offline) login. In that case going online really just triggers background sync and no visible action so it is not intrusive.

liwde commented:
@sleidig I agree and added this other behavior. I also added batching to the non-live sync, which seems to speed up the initial sync by factor 4 (4min->1min)

sleidig commented:
Another bug when offline:
1. try to log in with wrong credentials (correctly, the error message is displayed - we should change it's wording a little bit though as it doesn't have to be a first login. Maybe the same "Username or password incorrect!" that is shown when online and a second line "If you are trying to log in for the first time on this device, please make sure you have a working internet connection.")
2. log in with correct credentials. You are logged in but immediately end up in a login-logout-loop sending you forth and back between dashboard and login form

sleidig commented:
Everything else looks good!

liwde commented:
> we should change it's wording a little bit though as it doesn't have to be a first login

There was some flawed logic on my side, and fixing it required a bit more work. I introduced a separate SyncState `ABORTED`, as this better describes what happens to the initial sync when we are offline. This way, the UI can decide to show the correct error message.

sleidig commented:
> @sleidig I agree and added this other behavior. I also added batching to the non-live sync, which seems to speed up the initial sync by factor 4 (4min->1min)

is this batching actually the same as this issue: https://github.com/Aam-Digital/ndb-core/issues/175 ?

liwde commented:
> is this batching actually the same as this issue: #175 ?

Looks like it. At least I found that adding the batch option increased response sizes and sped up the sync. PouchDB default seems to be 25. And `_bulk_get` is used in some requests.
